# PR DAY 24

- Buatlah sebuah model analisis sentiment untuk sebuah film yang baru dirilis.
- Tugas dikumpulkan dalam bentuk PDF, yang di dalamnya meliputi:
  Step by step
  Jawaban atas pertanyaan:
1. Latar belakang pemilihan algoritma yang digunakan
2. Hasil evaluasi
3. Langkah-langkah untuk meningkatkan akurasi model
4. Link repository github (code, dataset, dan model)

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import re
import requests

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [20]:
data = pd.read_excel(r'/content/drive/MyDrive/Machine learning AI batch 2/DAY24/assignment/review_172days.xlsx')
data.head()

Text_Tweet Sentiment
0  172 days bener bener ya definisi abis di naiki...    Netral
1  Udah siap dibikin nangis nonton cerita Amer &a...    Netral
2  172 DAYS udah tayang di Netflix! Mengangkat ki...    Netral
3  AAAAAA SIAPA BUAT CERITA 172 DAYS NI https://t...   Positif
4                       emosi nya I tgk 172 days ni     Netral

## Text Processing

### Cleaning Text and Lower Case

In [5]:
def cleaning_text(text):
    # remove url
    url_pattern = re.compile(r'https?://\S+|www\.\S+')
    text = url_pattern.sub(r'', text)

    # remove hashtags
    # only removing the hash # sign from the word
    text = re.sub(r'#', '', text)

    # remove mention handle user (@)
    text = re.sub(r'@[\w]*', ' ', text)

    # remove emojis
    emoji_pattern = re.compile(
        '['
        '\U0001F600-\U0001F64F'  # emoticons
        '\U0001F300-\U0001F5FF'  # symbols & pictographs
        '\U0001F680-\U0001F6FF'  # transport & map symbols
        '\U0001F700-\U0001F77F'  # alchemical symbols
        '\U0001F780-\U0001F7FF'  # Geometric Shapes Extended
        '\U0001F800-\U0001F8FF'  # Supplemental Arrows-C
        '\U0001F900-\U0001F9FF'  # Supplemental Symbols and Pictographs
        '\U0001FA00-\U0001FA6F'  # Chess Symbols
        '\U0001FA70-\U0001FAFF'  # Symbols and Pictographs Extended-A
        '\U00002702-\U000027B0'  # Dingbats
        '\U000024C2-\U0001F251'
        ']+',
        flags=re.UNICODE
    )
    text = emoji_pattern.sub(r'', text)

    # remove punctuation
    punctuations = '''!()-[]{};:'"\,<>./?@#$%^&*_~'''
    for x in text.lower():
        if x in punctuations:
            text = text.replace(x, " ")

    # remove extra whitespace
    text = ' '.join(text.split())

    # lowercase
    text = text.lower()
    return text

### Remove Stopword

In [6]:
import nltk
from nltk.corpus import stopwords
nltk.download('stopwords')
nltk.download('punkt')

# CONSTRUCT STOPWORDS
rama_stopword = "https://raw.githubusercontent.com/ramaprakoso/analisis-sentimen/master/kamus/stopword.txt"
yutomo_stopword = "https://raw.githubusercontent.com/yasirutomo/python-sentianalysis-id/master/data/feature_list/stopwordsID.txt"
fpmipa_stopword = "https://raw.githubusercontent.com/onlyphantom/elangdev/master/elang/word2vec/utils/stopwords-list/fpmipa-stopwords.txt"
sastrawi_stopword = "https://raw.githubusercontent.com/onlyphantom/elangdev/master/elang/word2vec/utils/stopwords-list/sastrawi-stopwords.txt"
aliakbar_stopword = "https://raw.githubusercontent.com/onlyphantom/elangdev/master/elang/word2vec/utils/stopwords-list/aliakbars-bilp.txt"
pebahasa_stopword = "https://raw.githubusercontent.com/onlyphantom/elangdev/master/elang/word2vec/utils/stopwords-list/pebbie-pebahasa.txt"
elang_stopword = "https://raw.githubusercontent.com/onlyphantom/elangdev/master/elang/word2vec/utils/stopwords-id.txt"
nltk_stopword = stopwords.words('indonesian')

# create path url for each stopword
path_stopwords = [rama_stopword, yutomo_stopword, fpmipa_stopword, sastrawi_stopword,
                  aliakbar_stopword, pebahasa_stopword, elang_stopword]

# combine stopwords
stopwords_l = nltk_stopword
for path in path_stopwords:
    response = requests.get(path)
    stopwords_l += response.text.split('\n')

custom_st = '''
yg yang dgn ane smpai bgt gua gwa si tu ama utk udh btw
ntar lol ttg emg aj aja tll sy sih kalo nya trsa mnrt nih
ma dr ajaa tp akan bs bikin kta pas pdahl bnyak guys abis tnx
bang banget nang mas amat bangettt tjoy hemm haha sllu hrs lanjut
bgtu sbnrnya trjadi bgtu pdhl sm plg skrg
'''

# create dictionary with unique stopword
st_words = set(stopwords_l)
custom_stopword = set(custom_st.split())

# result stopwords
stop_words = st_words | custom_stopword
print(f'Stopwords: {list(stop_words)[:5]}')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


Stopwords: ['seingat', 'mana', 'jelaskan', 'tersebut', 'sebenarnya']


In [7]:
# remove stopwords
from nltk import word_tokenize, sent_tokenize

def remove_stopword(text, stop_words=stop_words):
    word_tokens = word_tokenize(text)
    filtered_sentence = [w for w in word_tokens if not w in stop_words]
    return ' '.join(filtered_sentence)

### Stemming / Lemmatization

In [9]:
!pip install sastrawi -q

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 209.7/209.7 kB 4.4 MB/s eta 0:00:00


In [10]:
# stemming and lemmatization
from Sastrawi.Stemmer.StemmerFactory import StemmerFactory

def stemming_and_lemmatization(text):
    factory = StemmerFactory()
    stemmer = factory.create_stemmer()
    return stemmer.stem(text)

###  Tokenization

In [12]:
# tokenization
def tokenize(text):
    return word_tokenize(text)

In [13]:
# example
text = 'Agak Laen ini emang agak lain, ya.. Bisa-bisanya jadi film Indonesia kedua terlaris 🙌🏼 #respect https://x.com/alnurulg/status/1761301921846140991?s=20'
print(f'Original text: \n{text}\n')

# cleaning text and lowercase
text = cleaning_text(text)
print(f'Cleaned text: \n{text}\n')

# remove stopwords
text = remove_stopword(text)
print(f'Removed stopword: \n{text}\n')

# stemming and lemmatization
text = stemming_and_lemmatization(text)
print(f'Stemmed and lemmatized: \n{text}\n')

# tokenization
text = tokenize(text)
print(f'Tokenized: \n{text}')

Original text: 
Agak Laen ini emang agak lain, ya.. Bisa-bisanya jadi film Indonesia kedua terlaris 🙌🏼 #respect https://x.com/alnurulg/status/1761301921846140991?s=20

Cleaned text: 
agak laen ini emang agak lain ya bisa bisanya jadi film indonesia kedua terlaris respect

Removed stopword: 
laen bisanya film indonesia terlaris respect

Stemmed and lemmatized: 
laen bisa film indonesia laris respect

Tokenized: 
['laen', 'bisa', 'film', 'indonesia', 'laris', 'respect']


In [16]:
# pipeline preprocess
def preprocess(text):
    # cleaning text and lowercase
    output = cleaning_text(text)

    # remove stopwords
    output = remove_stopword(output)

    # # stemming and lemmatization
    # output = stemming_and_lemmatization(output)

    # # tokenization
    # output = tokenize(output)

    return output

In [18]:
data

full_text Unnamed: 1
0    172 days bener bener ya definisi abis di naiki...     Netral
1    Udah siap dibikin nangis nonton cerita Amer &a...     Netral
2    172 DAYS udah tayang di Netflix! Mengangkat ki...     Netral
3    AAAAAA SIAPA BUAT CERITA 172 DAYS NI https://t...    Positif
4                         emosi nya I tgk 172 days ni      Netral
..                                                 ...        ...
135  @whosjaygf AWCH salting deh hari ini aku senen...     Netral
136               @sweetchcco ternyata 172 days baguss    Positif
137  kan baru sadar aku rumah di santri pilihan bun...     Netral
138              @sendalkukus gan coba nonton 172 days     Netral
139      @WatchmenID Bukan... 172 days (later) ya Min?     Netral

[140 rows x 2 columns]

In [23]:
preprocessed_data = data.copy()
preprocessed_data['Text Tweet'] = data['Text_Tweet'].map(preprocess)

In [24]:
preprocessed_data.tail()

Text_Tweet Sentiment  \
135  @whosjaygf AWCH salting deh hari ini aku senen...    Netral   
136               @sweetchcco ternyata 172 days baguss   Positif   
137  kan baru sadar aku rumah di santri pilihan bun...    Netral   
138              @sendalkukus gan coba nonton 172 days    Netral   
139      @WatchmenID Bukan... 172 days (later) ya Min?    Netral   

                                            Text Tweet  
135  awch salting seneng nonton film smkeluarga kar...  
136                                    172 days baguss  
137       sadar rumah santri pilihan bunda ck 172 days  
138                           gan coba nonton 172 days  
139                                 172 days later min

In [25]:
preprocessed_data['Text Tweet'][0]

'172 days bener bener definisi naikin tingginya dijatuhin sejatuh jatuhnya'

In [29]:
df = preprocessed_data[['Text Tweet', 'Sentiment']]

In [36]:
df.shape

(140, 2)

## LSTM
Dalam model sentiment analysis ini menggunakan model LSTM. LSTM mampu menangani sequence yang panjang dan kompleks, LSTM juga memiliki kemampuan untuk mengingat informasi jangka panjang.

In [30]:
import pandas as pd
from sklearn.model_selection import train_test_split
from keras.preprocessing.text import Tokenizer
from keras.preprocessing.sequence import pad_sequences
from keras.models import Sequential
from keras.layers import Embedding, LSTM, Dense

# Assuming the 'Text Tweet' column contains the text data and 'Sentiment' contains labels
texts = df['Text Tweet'].tolist()
labels = df['Sentiment'].tolist()

# Tokenize the text data
max_words = 10000  # Adjust based on your dataset size
tokenizer = Tokenizer(num_words=max_words)
tokenizer.fit_on_texts(texts)
sequences = tokenizer.texts_to_sequences(texts)

# Pad sequences to make them of equal length
max_sequence_length = 100  # Adjust based on your dataset and sequence length
data = pad_sequences(sequences, maxlen=max_sequence_length)

# Convert labels to one-hot encoding
labels = pd.get_dummies(labels)

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(data, labels, test_size=0.2, random_state=42)

# Build the LSTM model
model = Sequential()
model.add(Embedding(input_dim=max_words, output_dim=100, input_length=max_sequence_length))
model.add(LSTM(units=64, dropout=0.2, recurrent_dropout=0.2))
model.add(Dense(units=len(labels.columns), activation='softmax'))

model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Train the model
model.fit(X_train, y_train, epochs=100, batch_size=32, validation_data=(X_test, y_test))

# Evaluate the model
loss, accuracy = model.evaluate(X_test, y_test)
print(f"Test Loss: {loss:.4f}, Test Accuracy: {accuracy:.4f}")

Epoch 1/100
4/4 [==============================] - 5s 327ms/step - loss: 1.0877 - accuracy: 0.4911 - val_loss: 1.0827 - val_accuracy: 0.4643
Epoch 2/100
4/4 [==============================] - 1s 163ms/step - loss: 1.0282 - accuracy: 0.5893 - val_loss: 1.0675 - val_accuracy: 0.4643
Epoch 3/100
4/4 [==============================] - 1s 168ms/step - loss: 0.9611 - accuracy: 0.5893 - val_loss: 1.1703 - val_accuracy: 0.4643
Epoch 4/100
4/4 [==============================] - 1s 180ms/step - loss: 0.9295 - accuracy: 0.5893 - val_loss: 1.1187 - val_accuracy: 0.4643
Epoch 5/100
4/4 [==============================] - 1s 177ms/step - loss: 0.9028 - accuracy: 0.5893 - val_loss: 1.0615 - val_accuracy: 0.4643
Epoch 6/100
4/4 [==============================] - 1s 157ms/step - loss: 0.8792 - accuracy: 0.5982 - val_loss: 1.0495 - val_accuracy: 0.4643
Epoch 7/100
4/4 [==============================] - 1s 161ms/step - loss: 0.8500 - accuracy: 0.6250 - val_loss: 1.0687 - val_accuracy: 0.4643
Epoch 8/100
4

Model ini memilik akurasi 0.5, bisa dikatakan model masih kurang bagus dalam melakukan prediksi. Hal ini bisa jadi disebabkan oleh data yang digunakan masih terlalu sedikit, dalam processing ini data yang digunakan adalahh 140 data. Keterbatasan data dapat menyebabkan model tidak dapat memprediksi dengan presisi

In [31]:
# Save the model to a file
model.save('lstm_sentiment_model.h5')

# Optionally, save the tokenizer as well for later use during inference
import pickle

with open('tokenizer.pkl', 'wb') as tokenizer_file:
    pickle.dump(tokenizer, tokenizer_file)

/usr/local/lib/python3.10/dist-packages/keras/src/engine/training.py:3103: UserWarning: You are saving your model as an HDF5 file via `model.save()`. This file format is considered legacy. We recommend using instead the native Keras format, e.g. `model.save('my_model.keras')`.
  saving_api.save_model(


In [37]:
from keras.models import load_model
import pickle

# Load the model
loaded_model = load_model('lstm_sentiment_model.h5')

# Load the tokenizer
with open('tokenizer.pkl', 'rb') as tokenizer_file:
    loaded_tokenizer = pickle.load(tokenizer_file)

# New sentence for testing
test_sentence = 'nangis banget nontonnya'
new_sentence = preprocess(test_sentence)
# Tokenize and pad the new sentence
new_sequence = loaded_tokenizer.texts_to_sequences([new_sentence])
new_data = pad_sequences(new_sequence, maxlen=max_sequence_length)

# Predict sentiment for the new sentence
predictions = loaded_model.predict(new_data)

# Get the predicted sentiment label
predicted_sentiment_label = labels.columns[predictions.argmax(axis=1)[0]]
print(f"preprocessing: {new_sentence}")
print(f"new_sequence: {new_sequence}")
print(f"new_data: {new_data}")
print(f"Predicted Sentiment: {predicted_sentiment_label}")

1/1 [==============================] - 0s 253ms/step
preprocessing: nangis nontonnya
new_sequence: [[5]]
new_data: [[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
  0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 5]]
Predicted Sentiment: Positif


In [38]:
# New sentence for testing
test_sentence = 'cringe banget filmnya'
new_sentence = preprocess(test_sentence)
# Tokenize and pad the new sentence
new_sequence = loaded_tokenizer.texts_to_sequences([new_sentence])
new_data = pad_sequences(new_sequence, maxlen=max_sequence_length)

# Predict sentiment for the new sentence
predictions = loaded_model.predict(new_data)

# Get the predicted sentiment label
predicted_sentiment_label = labels.columns[predictions.argmax(axis=1)[0]]
print(f"preprocessing: {new_sentence}")
print(f"new_sequence: {new_sequence}")
print(f"new_data: {new_data}")
print(f"Predicted Sentiment: {predicted_sentiment_label}")

1/1 [==============================] - 0s 30ms/step
preprocessing: cringe filmnya
new_sequence: [[17, 352]]
new_data: [[  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0  17 352]]
Predicted Sentiment: Negatif


dilihat dari test tes di atas, model masih mampu memprediksi dengan benar

## Link github

Find code and data on github

https://github.com/wellyokt/ML2_DAY24.git